homeassistant/components/notify/xmpp.py

```diff
@@ -93,29 +118,217 @@ def __init__(self):
+            # also setting random filename for privacy
+            if data.get(ATTR_PATH):
+                # using given path as base for new filename. Don't guess type
+                filename = self.get_random_filename(data.get(ATTR_PATH))
+            else:
+                extension = mimetypes.guess_extension(
+                    result.headers['Content-Type']) or ".unknown"
+                _LOGGER.debug("Got %s extension", extension)
+                filename = self.get_random_filename(extension)
+
+            _LOGGER.info("Uploading file from URL, %s", filename)
+
+            url = await self['xep_0363'].upload_file(
+                filename, size=filesize, input_file=result.content,
+                content_type=result.headers['Content-Type'], timeout=timeout)
+
+            return url
+
+        async def upload_file_from_path(self, path, timeout=None):
+            """Upload a file from a local file path via XEP_0363."""
+            _LOGGER.info('Uploading file from path, %s ...', path)
+
+            with open(path, 'rb') as upfile:
+                _LOGGER.debug("Reading file %s", path)
+                input_file = upfile.read()
+            filesize = len(input_file)
+            _LOGGER.debug("Filesize is %s bytes", filesize)
+
+            content_type = mimetypes.guess_type(path)[0]
+            if content_type is None:
+                content_type = DEFAULT_CONTENT_TYPE
+            _LOGGER.debug("Content type is %s", content_type)
+
+            # set random filename for privacy
+            filename = self.get_random_filename(data.get(ATTR_PATH))
+            _LOGGER.debug("Uploading file with random filename %s", filename)
+
+            url = await self['xep_0363'].upload_file(
+                filename, size=filesize, input_file=input_file,
+                content_type=content_type, timeout=timeout)
+
+            return url
+
+        def send_text_message(self):
+            """Send a text only message to a room or a recipient."""
+            try:
+                if room:
+                    _LOGGER.debug("Joining room %s", room)
+                    self.plugin['xep_0045'].join_muc(room, sender, wait=True)
+                    self.send_message(
+                        mto=room, mbody=message, mtype='groupchat')
+                else:
+                    _LOGGER.debug("Sending message to %s", recipient)
+                    self.send_message(
+                        mto=recipient, mbody=message, mtype='chat')
+            except (IqError, IqTimeout, XMPPError) as ex:
+                _LOGGER.error("Could not send text message %s", ex)
+            except NotConnectedError as ex:
+                _LOGGER.error("Connection error %s", ex)
+
+        # pylint: disable=no-self-use
+        def get_random_filename(self, filename):
+            """Return a random filename, leaving the extension intact."""
+            if '.' in filename:
+                extension = filename.split('.')[-1]
```

Comment:
This could fail for strings like `/home/user/data/../file` which is maybe not completely unlikely with templates in use(?)

Reply:
yes, will fix.